source/adapters/opencl/platform.cpp

```diff
@@ -83,10 +66,38 @@ UR_APIEXPORT ur_result_t UR_APICALL
 urPlatformGet(ur_adapter_handle_t *, uint32_t, uint32_t NumEntries,
               ur_platform_handle_t *phPlatforms, uint32_t *pNumPlatforms) {
 
-  cl_int Result =
-      clGetPlatformIDs(cl_adapter::cast<cl_uint>(NumEntries),
-                       cl_adapter::cast<cl_platform_id *>(phPlatforms),
-                       cl_adapter::cast<cl_uint *>(pNumPlatforms));
+  static std::vector<std::unique_ptr<ur_platform_handle_t_>> URPlatforms;
+  static std::once_flag InitFlag;
+  static uint32_t NumPlatforms = 0;
+  cl_int Result = CL_SUCCESS;
+
+  std::call_once(
+      InitFlag,
+      [](cl_int &Res) {
+        Res = clGetPlatformIDs(0, nullptr, &NumPlatforms);
+        if (Res != CL_SUCCESS) {
+          return Res;
+        }
+        std::vector<cl_platform_id> CLPlatforms(NumPlatforms);
+        Res = clGetPlatformIDs(static_cast<cl_uint>(NumPlatforms),
+                               CLPlatforms.data(), nullptr);
+        if (Res != CL_SUCCESS) {
+          return Res;
+        }
+        try {
+          for (uint32_t i = 0; i < NumPlatforms; i++) {
+            auto URPlatform =
+                std::make_unique<ur_platform_handle_t_>(CLPlatforms[i]);
+            URPlatforms.emplace_back(URPlatform.release());
+          }
+        } catch (std::bad_alloc &) {
+          return CL_OUT_OF_RESOURCES;
+        } catch (...) {
+          return CL_INVALID_PLATFORM;
+        }
+        return Res;
+      },
+      Result);
```

Comment:
Do we already have a `call_once` elsewhere? Might be better to combine them if we do.

Reply:
I already refactored this inside adapter so no need for `call_once`.